File: Diagrammatize.Design.Pattern/code/cxx/Mediator/Sample/ColleagueCheckbox.cpp
#include "ColleagueCheckbox.hpp"

#include "Mediator.hpp"

ColleagueCheckbox::ColleagueCheckbox(const std::string& caption, bool checked, QWidget *parent)
    : QRadioButton(caption.c_str(), parent) {

    setChecked(checked);
    connect(this, SIGNAL(toggled(bool)), this, SLOT(itemStateChanged(bool)));
}

ColleagueCheckbox::~ColleagueCheckbox() {}

void ColleagueCheckbox::setMediator(Mediator* mediator_) {    // 保存Mediator
    mediator = mediator_;
}

void ColleagueCheckbox::setColleagueEnabled(bool enabled) {   // Mediator下达启用/禁用的指示
    setEnabled(enabled);
}

void ColleagueCheckbox::itemStateChanged(bool on) {             // 当状态发生变化时通知Mediator
    if (on) mediator->colleagueChanged();
}

File: Diagrammatize.Design.Patterns/code/python/Mediator/A1/ColleagueCheckbox.py
import tkinter as tk
from Colleague import Colleague

class ColleagueCheckbox(tk.Radiobutton, Colleague):
    def __init__(self, parent, caption, var:tk.IntVar, value:int, font=None):
        self.mediator = None
        self.var = var
        self.value = value
        tk.Radiobutton.__init__(self, parent, text=caption, variable=var, value=value, command=self.itemStateChanged, font=font)

    def setMediator(self, mediator):
        self.mediator = mediator

    def itemStateChanged(self):
        if self.mediator:
            self.mediator.colleagueChanged()
        else:
            print("itemStateChanged, but mediator is None")

    def setColleagueEnabled(self, enabled):
        if enabled:
            self.config(state=tk.NORMAL)
        else:
            self.config(state=tk.DISABLED)

    def getState(self):
        if self.value == self.var.get():
            return True
        else:
            return False

def main():
    root = tk.Tk()
    var = tk.IntVar()
    var.set(0)
    checkbox = ColleagueCheckbox(root, "hello", var, 1)
    checkbox.pack()
    checkbox.setColleagueEnabled(True)
    print("checkbox.getState:", checkbox.getState())
    root.mainloop()

if __name__ == "__main__":
    main()

